core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2826,75 +2922,312 @@ public void cleanUp() {
+            LOGGER.log(SEVERE, "Failed to shut down Jenkins load thread pool", e);
+            // safe to ignore and continue for this one
+        } catch (Throwable e) {
+            LOGGER.log(SEVERE, "Failed to shut down Jenkins load thread pool", e);
+            // save for later
+            errors.add(e);
+        }
+    }
 
-        threadPoolForLoad.shutdown();
-        for (Future<?> f : pending)
+    private void _cleanUpAwaitDisconnects(List<Throwable> errors, Set<Future<?>> pending) {
+        if (!pending.isEmpty()) {
+            LOGGER.log(Level.INFO, "Waiting for node disconnection completion");
+        }
+        for (Future<?> f : pending) {
             try {
                 f.get(10, TimeUnit.SECONDS);    // if clean up operation didn't complete in time, we fail the test
             } catch (InterruptedException e) {
                 Thread.currentThread().interrupt();
                 break;  // someone wants us to die now. quick!
             } catch (ExecutionException e) {
-                LOGGER.log(Level.WARNING, "Failed to shut down properly",e);
+                LOGGER.log(Level.WARNING, "Failed to shut down remote computer connection cleanly", e);
             } catch (TimeoutException e) {
-                LOGGER.log(Level.WARNING, "Failed to shut down properly",e);
+                LOGGER.log(Level.WARNING, "Failed to shut down remote computer connection within 10 seconds", e);
+            } catch (OutOfMemoryError e) {
+                // we should just propagate this, no point trying to log
+                throw e;
+            } catch (LinkageError e) {
+                LOGGER.log(Level.WARNING, "Failed to shut down remote computer connection", e);
+            } catch (Throwable e) {
+                LOGGER.log(Level.SEVERE, "Unexpected error while waiting for remote computer connection disconnect", e);
+                errors.add(e);
             }
+        }
+    }
 
-        PluginServletFilter.cleanUp();
-
-        LogFactory.releaseAll();
+    private void _cleanUpPluginServletFilters(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Stopping filters");
+        try {
+            PluginServletFilter.cleanUp();
+        } catch (OutOfMemoryError e) {
+            // we should just propagate this, no point trying to log
+            throw e;
+        } catch (LinkageError e) {
+            LOGGER.log(SEVERE, "Failed to stop filters", e);
+            // safe to ignore and continue for this one
+        } catch (Throwable e) {
+            LOGGER.log(SEVERE, "Failed to stop filters", e);
+            // save for later
+            errors.add(e);
+        }
+    }
 
-        theInstance = null;
-        if (JenkinsJVM.isJenkinsJVM()) {
-            JenkinsJVMAccess._setJenkinsJVM(oldJenkinsJVM);
+    private void _cleanUpReleaseAllLoggers(List<Throwable> errors) {
+        LOGGER.log(Level.FINE, "Releasing all loggers");
+        try {
+            LogFactory.releaseAll();
```

Comment:
Not part of this change - but this seems wrong as we don't ship with commons-logging but use the SLF4J bridge which uses jul.